ompi/mca/coll/han/coll_han.h

```diff
@@ -486,4 +508,18 @@ ompi_coll_han_reorder_gather(const void *sbuf,
                              struct ompi_communicator_t *comm,
                              int * topo);
 
+
+static inline struct mca_smsc_endpoint_t *mca_coll_han_get_smsc_endpoint (struct ompi_proc_t *proc) {
+    if (OPAL_UNLIKELY(NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC])) {
+        /* LAR TODO: should we make a lock for this? */
+        /* OPAL_THREAD_LOCK(&XXX); */
+        if (NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC]) {
+            proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC] = mca_smsc->get_endpoint(&proc->super);
+        }
+        /* OPAL_THREAD_UNLOCK(&XXX); */
```

Comment:
I think there should be a lock with a double check. Something like this:

```suggestion
        if (NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC]) {
            OPAL_THREAD_LOCK(&XXX);
            if (NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC]) {
                proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC] = mca_smsc->get_endpoint(&proc->super);
            }
            OPAL_THREAD_UNLOCK(&XXX);
        }
```

Reply:
I stole this code from [bml](https://github.com/open-mpi/ompi/blob/42e1cd871ccde6f625c01562818cf2ba02ac7faf/ompi/mca/bml/base/base.h#L66) which just uses a bml lock.  I suppose I could do the same for an smsc lock, I'll just need to create one.

Reply:
@devreal I implemented this with a new lock scoped to the han module.

Although currently han is the only module to use the smsc endpoint, in the future this might not be true.  I feel like the lock should really be in ompi_proc_t and could apply to all get_endpoint calls, but this pattern isn't followed elsewhere so I'm hesitant to introduce it, so here we are with a new han lock.

Thoughts?